没有 IPv6 Scope ID 会导致异常

![屏幕截图 2024-05-30 142102](https://github.com/andreikop/python-ws-discovery/assets/5992155/fd12d7c5-0b10-4b18-981b-547669fb258d)

I have a similar error. With this stack trace.

```
    wsd.start()
venv/lib64/python3.12/site-packages/wsdiscovery/threaded.py:404: in start
    self._startThreads()
venv/lib64/python3.12/site-packages/wsdiscovery/threaded.py:380: in _startThreads
    self._addrsMonitorThread_v6 = AddressMonitorThread(self, socket.AF_INET6)
venv/lib64/python3.12/site-packages/wsdiscovery/threaded.py:52: in __init__
    self._updateAddrs()
venv/lib64/python3.12/site-packages/wsdiscovery/threaded.py:64: in _updateAddrs
    self._wsd._networkAddressAdded(addr)
venv/lib64/python3.12/site-packages/wsdiscovery/discovery.py:101: in _networkAddressAdded
    self.addSourceAddr(addr)
venv/lib64/python3.12/site-packages/wsdiscovery/threaded.py:417: in addSourceAddr
    self._networkingThread_v6.addSourceAddr(addr)
venv/lib64/python3.12/site-packages/wsdiscovery/threaded.py:141: in addSourceAddr
    self._multiInSocket.setsockopt(self._get_ip_proto(), self._get_ip_join(), self._makeMreq(addr))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <NetworkingThreadIPv6(Thread-2, started daemon 140235888126496)>
addr = IPv6Address('fe80::42:6bff:fea0:1b11%docker0')

    def _makeMreq(self, addr):
>       return struct.pack("=16si", socket.inet_pton(socket.AF_INET6, MULTICAST_IPV6_ADDRESS), int(addr.scope_id))
E       ValueError: invalid literal for int() with base 10: 'docker0'
```

--- a/README.md
+++ b/README.md
@@ -51,7 +51,7 @@ environment and set up a development environment. For example:
 ```
 $ python3 -m venv venv
 $ . venv/bin/activate
-(venv) $ pip3 install -e
+(venv) $ pip3 install -e .
 ```
 Then you can call `wsdiscover` and `wspublish` from within this virtual
 environment. Any code changes to the package will be available and testable

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,2 +1,2 @@
-netifaces-plus
+ifaddr
 click

--- a/setup.py
+++ b/setup.py
@@ -40,8 +40,8 @@ setup(name='WSDiscovery',
             'Topic :: Communications'
       ],
       packages=['wsdiscovery', 'wsdiscovery.actions'],
-      setup_requires=['netifaces-plus', 'click'],
-      install_requires=['netifaces-plus', 'click'],
+      setup_requires=['ifaddr', 'click'],
+      install_requires=['ifaddr', 'click'],
       tests_require = ['pytest', 'mock'],
       entry_points = {
          'console_scripts': [

--- a/wsdiscovery/threaded.py
+++ b/wsdiscovery/threaded.py
@@ -308,6 +308,7 @@ class NetworkingThreadIPv4(NetworkingThread):
         super().__init__(observer)
 
     def _makeMreq(self, addr):
+        "IPvv multicast group join/leave request"
         return struct.pack("4s4s", socket.inet_aton(MULTICAST_IPV4_ADDRESS), addr.packed)
 
     def _get_inet(self):
@@ -334,6 +335,7 @@ class NetworkingThreadIPv6(NetworkingThread):
         super().__init__(observer)
 
     def _makeMreq(self, addr):
+        "IPv6 multicast group join/leave request"
         return struct.pack("=16si", socket.inet_pton(socket.AF_INET6, MULTICAST_IPV6_ADDRESS), int(addr.scope_id))
 
     def _get_inet(self):

--- a/wsdiscovery/util.py
+++ b/wsdiscovery/util.py
@@ -1,15 +1,20 @@
 """Various utilities used by different parts of the package."""
 
 import io
-import ipaddress
 import string
 import random
-import netifaces
+import logging
+import ipaddress
+import socket
 from xml.dom import minidom
+import ifaddr
+
 from .scope import Scope
 from .uri import URI
 from .namespaces import NS_ADDRESSING, NS_DISCOVERY, NS_SOAPENV
 from .qname import QName
+
+logger = logging.getLogger("util")
 
 
 def createSkelSoapMessage(soapAction):
@@ -245,17 +250,29 @@ def getQNameFromValue(value, node):
 
 
 def _getNetworkAddrs(protocol_version):
-    result = []
+    addrs = []
+    ifaces = ifaddr.get_adapters()
 
-    for if_name in netifaces.interfaces():
-        iface_info = netifaces.ifaddresses(if_name)
-        if protocol_version in iface_info:
-            for addrDict in iface_info[protocol_version]:
-                addr = addrDict['addr']
-                ip_address = ipaddress.ip_address(addr)
-                if not ip_address.is_loopback:
-                    result.append(ip_address)
-    return result
+    # ifaddr library only returns IPv4 and IPv6 addresses
+    if protocol_version == socket.AF_INET:
+        for iface in ifaces:
+            for ip in iface.ips:
+                if type(ip.ip) is string:
+                    ip_address = ipaddress.ip_address(ip.ip)
+                    if not ip_address.is_loopback:
+                        addrs.append(ip_address)
+    elif protocol_version == socket.AF_INET6:
+        for iface in ifaces:
+            for ip in iface.ips:
+                if type(ip.ip) is tuple:
+                    ip_address = ipaddress.ip_address(f"{ip.ip[0]}%{ip.ip[2]}")
+                    if not ip_address.is_loopback:
+                        addrs.append(ip_address)
+    else:
+        logger.warning(f"requested protocol version ({protocol_version}) is not"
+                       f" IPv4 ({socket.AF_INET}) or IPv6 ({socket.AF_INET6})")
+
+    return addrs
 
 
 def _generateInstanceId():

